lib/marked.js

```diff
@@ -454,111 +843,221 @@ Lexer.prototype.token = function(src, top) {
 };
 
 /**
- * Inline-Level Grammar
+ * Expose
+ *
  */
+Lexer.rules = block;
 
-var inline = {
-  escape: /^\\([\\`*{}\[\]()#+\-.!_>])/,
-  autolink: /^<(scheme:[^\s\x00-\x1f<>]*|email)>/,
-  url: noop,
-  tag: /^<!--[\s\S]*?-->|^<\/?[a-zA-Z0-9\-]+(?:"[^"]*"|'[^']*'|\s[^<'">\/\s]*)*?\/?>/,
-  link: /^!?\[(inside)\]\(href\)/,
-  reflink: /^!?\[(inside)\]\s*\[([^\]]*)\]/,
-  nolink: /^!?\[((?:\[[^\[\]]*\]|\\[\[\]]|[^\[\]])*)\]/,
-  strong: /^__([\s\S]+?)__(?!_)|^\*\*([\s\S]+?)\*\*(?!\*)/,
-  em: /^_([^\s_](?:[^_]|__)+?[^\s_])_\b|^\*((?:\*\*|[^*])+?)\*(?!\*)/,
-  code: /^(`+)\s*([\s\S]*?[^`]?)\s*\1(?!`)/,
-  br: /^ {2,}\n(?!\s*$)/,
-  del: noop,
-  text: /^[\s\S]+?(?=[\\<!\[`*]|\b_| {2,}\n|$)/
-};
-
-inline._scheme = /[a-zA-Z][a-zA-Z0-9+.-]{1,31}/;
-inline._email = /[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+(@)[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)+(?![-_])/;
-
-inline.autolink = edit(inline.autolink)
-  .replace('scheme', inline._scheme)
-  .replace('email', inline._email)
-  .getRegex()
-
-inline._inside = /(?:\[[^\[\]]*\]|\\[\[\]]|[^\[\]]|\](?=[^\[]*\]))*/;
-inline._href = /\s*<?([\s\S]*?)>?(?:\s+['"]([\s\S]*?)['"])?\s*/;
-
-inline.link = edit(inline.link)
-  .replace('inside', inline._inside)
-  .replace('href', inline._href)
-  .getRegex();
+/**
+ * Parsing & Compiling
+ */
 
-inline.reflink = edit(inline.reflink)
-  .replace('inside', inline._inside)
-  .getRegex();
+function Parser(options) {
+  this.tokens = [];
+  this.token = null;
```

Comment:
Is having it be `null` necessary? Given that this becomes and object with members might be easier for a new person to understand if they saw something like this (or make a Token object):

```js
{
  type: '',
  ...
}
```

Was going to make a more complete example; however, it appears the Token is just a random set of members...like an undefined type. You can see it in the `switch` inside `Parser.prototype.tok()` ~ line 930

Reply:
Maybe something more akin to this from traditional OO languages...using Swift highlighting, but not the syntax, strictly speaking.

```swift
abstract class Token {}

class Heading {}

// then the switch could be something like
switch (typeof this.token) {
  case 'Heading':
...
```

But, if the `type` member is the only thing in common between all the different paths for Token...then there's really no reason for the `type` member or Token. ??

Reply:
I wonder if there are more "hidden" objects inside of here.